pkg/dns/service.go

```diff
@@ -288,20 +288,30 @@ func (s *Service) PatchTargets(ctx context.Context, targets, hosts []string, clu
 	return nil
 }
 
-// this is temporary and will be replaced in the future by CRD resources
-type zone struct {
-	v1.DNSZone
-	RootDomain string
-	Default    bool
-}
+// GetManagedZone returns the most suitable ManagedZone to publish the given traffic resource into.
+// Currently, this just returns the first ManagedZone found in the traffic resources own namespace, or if none is found,
+// it looks for the first one listed in the default ctrl namespace.
+func (s *Service) GetManagedZone(ctx context.Context, t traffic.Interface) (*v1alpha1.ManagedZone, error) {
```

Comment:
I'm thinking about how to work this into Gateways, and wondering if the current implementation is what we need for both Ingress and Gateways.
Both will have potentially multiple hosts (Ingress.spec.rules and Gateway.spec.listeners), so more than 1 MZ may be required.
I think for an initial implementation, what's here would work for both, with the limitation that you have to keep all hosts in the same 'zone'.

Reply:
Yeah we might want to adjust this to be GetManagedZones? or have this take the specific host from the listener/spec?

Reply:
Per host/listener is more useful than getting back a list of MZs.
I'm thinking a function that takes a list of hosts (strings, so it works for Ingress & Gateways), and returns a map of hosts to MZ.
A per host lookup function may be useful too, but it might be too easy to overuse it in a loop and hit a performance penalty as each loop may be doing a lot of the same data lookup and mapping.

Reply:
No need to change this right now though btw.
Let's refactor it to satisfy the Gateway use case in #84

Reply:
Yes, this will need updated once it's merged to work with Gateways. My understanding is the logic will be slightly different, requiring a lookup of managed zones based on the host(s) in the gateway spec.